main/src/main/scala/sbt/ExecuteProgress2.scala

```diff
@@ -0,0 +1,79 @@
+/*
+ * sbt
+ * Copyright 2023, Scala center
+ * Copyright 2011 - 2022, Lightbend, Inc.
+ * Copyright 2008 - 2010, Mark Harrah
+ * Licensed under Apache License 2.0 (see LICENSE)
+ */
+
+package sbt
+import sbt.internal.util.RMap
+
+/**
+ * Tracks command execution progress. In addition to ExecuteProgress, this interface
+ * adds command start and end events, and gives access to the sbt.State at the beginning
+ * and end of each command.
+ */
+trait ExecuteProgress2 extends ExecuteProgress[Task] {
```

Comment:
Since the original `ExecuteProgress` is abstracted over `Task`, it might be useful to continue with the tradition too, in case we want ExecuteProgress3 etc in the future.

Reply:
My reasoning was that the type we exposed in sbt.Keys was `TaskProgress`, so the type parameters was fixed, but I see no problem to add it here.

Reply:
I added an attempt to keep the type parameter in https://github.com/sbt/sbt/pull/7350/commits/18d12c2dbf61b249683711bfd5a3c8a4deac16ff. However, if the reasoning below resonates with you, I would prefer to keep the current version:

- we're hitting limitations in higher-kinded support in Scala, which prevents us to express directly something like `AttributeKey[ExecuteProgress2[Task]]`. The work around is to have another type that instantiates the higher-kinded type (probably another instance of https://github.com/scala/bug/issues/2915), so we end up still having a fixed `ExecuteProgress2[Task]` (similar to today's situation for `progressReports`)
- this abstraction only makes sense in the `tasks` subprojects. In the sbt `main` project, everything is fixed at `Task`, including `EvaluateTaskConfig`. This is part of public interfaces (`progressReports` is also concrete, same for `EvaluateTask`) and there's no way to run the eval engine on anything else.
- to work around the bug we need to introduce a concrete type and use it in `sbt.Keys` which is structurally identical to `ExecuteProgress2[Task]` but it's seen as a different type (a type alias hits the same compiler bug). I find this will be confusing to users and leaks a limitation in the type system in ways that won't give us any benefit.

Given the above, I would choose the existing implementation, but have a look for yourself and let me know what you think.

Once we are happy with the implementation, I will add a scripted test.

Reply:
> we're hitting limitations in higher-kinded support in Scala

> Given the above, I would choose the existing implementation, but have a look for yourself and let me know what you think.

ok. Then the existing impl sounds good to me. Thanks for digging into it!

Reply:
Great, will add a scripted test and ping you back.